### ISSUE TITLE
JWT does not seem to authenticate in Python

### ISSUE THREAD

**Issue by JaviBonilla** [appwrite/appwrite#1251]
## 🐛 Bug Report

JWT does not seem to authenticate in Python

## Have you spent some time to check if this issue has been raised before?

Yes

## To Reproduce

```python
client = Client().set_endpoint(ENDPOINT).set_project(PROJECT).set_j_w_t(token)
account = Account(client)
account.get()
```

## Expected behavior

Autheticate the user

## Actual Behavior

`User (role: guest) missing scope (account)`

Before that, the `lower()` method must be removed in `client.py`

```python
def set_j_w_t(self, value):
     """Your secret JSON Web Token"""

     self._global_headers['x-appwrite-jwt'] = value #.lower()
     return self
```

Otherwise:

`Failed to verify JWT. JSON failed: Syntax error`

## Your Environment

- Appwrite 0.8
- Python 3.8
- Raspberry Pi 4
- Ubuntu Server 20.04.2 LTS

**Comment by JaviBonilla** [comment]
I can decode the token in Python, so the token is not the problem.

```python
import jwt

data = jwt.decode(token, SECRET_KEY, algorithms='HS256')
print(data)
```
```json
{"userId": "60acda002d991", "sessionId": "60bb47ba1d28e", "exp": 1622891198}
```

I'm testing this locally, outside Appwrite functions, any problem with that?

**Comment by eldadfux** [comment]
Work in progress here: https://github.com/appwrite/sdk-generator/pull/207

**Comment by TorstenDittmann** [comment]
Should be released now. Can you verify @JaviBonilla ? 🙏🏻

**Comment by JaviBonilla** [comment]
@TorstenDittmann, the `Failed to verify JWT. JSON failed: Syntax error` problem is solved, but I still get this error:

```
User (role: guest) missing scope (account)
```

With this code:

```python
client = Client().set_endpoint(server.ENDPOINT).set_project(server.PROJECT).set_jwt(token)
account = Account(client)
account.get()
```
I'm executing this otuside of Appwrite functions for testing purposes (Python code is executed in my laptop, Appwrite is in a Raspberry Pi), any problem with that? Below it's all the information in the API call:

```
Method: get
URL: https://192.168.1.30/v1/account
Params: {}
Data: {}
JSON: {}
Files: {}
Headers: {
  "content-type": "application/json",
  "x-sdk-version": "appwrite:python:0.3.0",
  "X-Appwrite-Response-Format": "0.8.0",
  "x-appwrite-project": "60aab01f8b0de",
  "x-appwrite-jwt": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOiI2MGFjZGEwMDJkOTkxIiwic2Vzc2lvbklkIjoiNjBiZTg5ODY2YTNmYSIsImV4cCI6MTYyMzEwMTQwOX0.Mmc2gKeE-OMneYraNx4h4S_i_IYYldHxXjiq0zt0VuY"
}
Verify: False
```

**Comment by eldadfux** [comment]
This would mean that authentication failed and user is identified as a guest user. In a working auth flow this would indicate one of two things:
1. JWT has expired (default expiry interval is 15 minutes)
2. The user session that was used to generate the JWT has been deleted before the JWT has been expired

**Comment by JaviBonilla** [comment]
Thanks @eldadfux,

To make sure I was doing it right, from my client app (Flutter), I got the token and just after that tried to validate it in the client. 

It works if Appwrite is in my localhost

```
Token generated: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOiI2MGJmNGRmYzJjN2Y0Iiwic2Vzc2lvbklkIjoiNjBiZjZiN2U4NGE2NyIsImV4cCI6MTYyMzE2MzY1OH0.xk0ZLewMIIKh61-GWlqU46B3IxQWVZ4YAc2DW_0I5L4
Token payload: {"userId": "60bf4dfc2c7f4", "sessionId": "60bf6b7e84a67", "exp": 1623163658}
Token validated, user account: {"$id":"60bf4dfc2c7f4","name":"Javier","registration":1623150076,"status":0,"passwordUpdate":1623150076,"email":"xxxxxx@gmail.com","emailVerification":true,"prefs":{}}

```

But it fails if Appwrite is in my Raspberry Pi with the same code:

```
Token generated: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOiI2MGFjZGEwMDJkOTkxIiwic2Vzc2lvbklkIjoiNjBiZjgwNzU2ZTIxYiIsImV4cCI6MTYyMzE2Mzk1MH0.CxaKH8FPCffDMqbzdUYs4pzzj1WIVHWkU6t4K2Py13o
Token payload: {"userId": "60acda002d991", "sessionId": "60bf80756e21b", "exp": 1623163950}
Exception: User (role: guest) missing scope (account)
```

Is there something that I have to take into account when Appwrite is in another host or on a Raspberry Pi when setting the JWT in the Appwrite client?

**Comment by eldadfux** [comment]
You can't move the JWT between different Appwrite instances, as the user must be signed in the Appwrite instance you're validating against.

**Comment by JaviBonilla** [comment]
They are independent executions, the token is different and generated on each Appwrite instance. Same client but connected each time to a different server.

Some configuration is probably wrong with my Raspberry pi Appwrite server, the token is generated but not validated.

**Comment by JakeAi** [comment]
@eldadfux I seem to be having the same issue with NodeJS rather than python.
```TS
const user = await this.appwriteService.setJwt(jwt).account.get();

export class AppwriteService {
  constructor(private configService: ConfigService) {
    const client = new Client();
    client
      .setEndpoint('url') // Your API Endpoint
      .setProject('project');
    this._client = client;
  }

  public get account(): Account {
    return new Account(this._client);  
  }

  public setJwt(jwt: string) {
    this._client = this._client.setJWT(jwt);
    return this;
  }
}
```
```
Client {
  endpoint: 'myurlishere',
  headers: {
    'content-type': '',
    'x-sdk-version': 'appwrite:nodejs:2.3.0',
    'X-Appwrite-Response-Format': '0.8.0',
    'x-appwrite-project': 'myprojectcodeishere',
    'x-appwrite-jwt': 'myvalidjsontokenisactuallyhereaswell'
  },
  selfSigned: false
}
```
when i console log client, it looks like everything is actually set correctly and after 15mins, the token expires, it actually tells me the token expired
```
  code: 401,
  response: {
    message: 'Failed to verify JWT. Invalid token: Expired',
    code: 401,
    version: '0.8.0'
  }

```

**Comment by eldadfux** [comment]
> @eldadfux I seem to be having the same issue with NodeJS rather than python.
> 
> ```ts
> const user = await this.appwriteService.setJwt(jwt).account.get();
> 
> export class AppwriteService {
>   constructor(private configService: ConfigService) {
>     const client = new Client();
>     client
>       .setEndpoint('url') // Your API Endpoint
>       .setProject('project');
>     this._client = client;
>   }
> 
>   public get account(): Account {
>     return new Account(this._client);  
>   }
> 
>   public setJwt(jwt: string) {
>     this._client = this._client.setJWT(jwt);
>     return this;
>   }
> }
> ```
> 
> ```
> Client {
>   endpoint: 'myurlishere',
>   headers: {
>     'content-type': '',
>     'x-sdk-version': 'appwrite:nodejs:2.3.0',
>     'X-Appwrite-Response-Format': '0.8.0',
>     'x-appwrite-project': 'myprojectcodeishere',
>     'x-appwrite-jwt': 'myvalidjsontokenisactuallyhereaswell'
>   },
>   selfSigned: false
> }
> ```
> 
> ```
> code: 401,
>   response: {
>     message: 'User (role: guest) missing scope (account)',
>     code: 401,
>     version: '0.8.0'
>   }
> ```
> 
> when i console log client, it looks like everything is actually set correctly and after 15mins, the token expires, it actually tells me the token expired
> 
> ```
>   code: 401,
>   response: {
>     message: 'Failed to verify JWT. Invalid token: Expired',
>     code: 401,
>     version: '0.8.0'
>   }
> ```

Could you also share the client side code you use to generate the JWT?

**Comment by PineappleIOnic** [comment]
I've created the most barebones way of creating a JWT token. Could you please try it and let me know if it works for you? Just a note. This uses the CDN Version of Appwrite's JS Library

```js
```

**Comment by eldadfux** [comment]
Another thing I think we should mention again about this issue. We are aware of a cross-SDK bug causing the SDK to lowercase the JWT value before sending it to the Appwrite API - which obviously results in an authentication failure. This was fixed last week - make sure you run the latest version of your Appwrite SDK.

**Comment by JakeAi** [comment]
This is the raw headers that were pulled from wireshark from my nodejs api.
```
GET /v1/account HTTP/1.1
Accept: application/json, text/plain, */*
x-sdk-version: appwrite:nodejs:2.3.0
X-Appwrite-Response-Format: 0.8.0
x-appwrite-project: 60ba41c8937c5
x-appwrite-jwt: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOiI2MGJlNzdiNjE1Nzk2Iiwic2Vzc2lvbklkIjoiNjBjYjZlODE1ZDVmNyIsImV4cCI6MTYyMzk2Njc3Mn0.WSRsG9lfKnLAmecYh1CVG9JLsaZGhwlbW0uqEaO4Zrs
Content-Type: application/json
User-Agent: axios/0.21.1
Host: my-endpoint.com
Connection: close

```

**Comment by JakeAi** [comment]
> I've created the most barebones way of creating a JWT token. Could you please try it and let me know if it works for you? Just a note. This uses the CDN Version of Appwrite's JS Library
> 
> ```js
> const appwrite = new Appwrite.Appwrite();
> 
> const email = 'amadeus@test.co.uk';
> const password = 'testtesttest';
> 
> appwrite
>     .setEndpoint('http://localhost/v1') // Set only when using self-hosted solution
>     .setProject('PROJECTID');
> 
> appwrite.account.createSession(email, password).then(function (response) {
>     appwrite.account.createJWT().then(function (response) {
>         alert(response.jwt);
>     })
> })
> ```
> 
> Make sure to update the Project ID, URL, Email and Password to your values 🙂

`eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOiI2MGJlNzdiNjE1Nzk2Iiwic2Vzc2lvbklkIjoiNjBjZTk0NDk0NzMzYiIsImV4cCI6MTYyNDE1MjAxM30.rn04HB9d03MFtMttROSPZ7Z4yJD6PEVcNepXcAir1rQ`

**Comment by JaviBonilla** [comment]
I’m having exactly the same issue, I’m almost sure it is not related to the client code, also it is not easy to replicate.

Currently I have two Appwrite servers installed for testing with the same configuration, the one at my laptop is working fine, after reinstalling Appwrite, not sure why this solved the problem,  because I had this issue before.

https://github.com/appwrite/sdk-for-python/issues/31

The other in my Raspberry Pi has this problem, tried to reinstall but it didn’t solve it. My client code is the same, of course the connection configuration, endpoint, project id, etc is different.

Could more information be printed in the Appwrite logs from this php code to try to figure out what is going on?

**Comment by eldadfux** [comment]
@JaviBonilla can you try to add logs to this file in your running Appwrite?

https://github.com/appwrite/appwrite/blob/master/app/init.php#L442-L463

For code changes to take effect you need to restart your Appwrite container

**Comment by JaviBonilla** [comment]
@eldadfux @JakeAi I created a new user and now the JWT tokens are working for this user! I accessed the database, I don't have much knwoledge about the internal structure, so maybe I'm wrong, but there is some info storer for the previous user (_60acda002d991_) than I cannot find for the new user (_60cfa90a4fe23_) in the _properties_ and _relationships_ tables.

![token](https://user-images.githubusercontent.com/20316179/122688262-9b83fb00-d21b-11eb-8d20-868f16483804.png)

![image](https://user-images.githubusercontent.com/20316179/122688441-75ab2600-d21c-11eb-9286-3635bace07ab.png)

**Comment by JakeAi** [comment]
Making a new user does work for me too. So you can't use a root account?

**Comment by JaviBonilla** [comment]
@JakeAi Not sure, but there is a comment about the root user in the code, https://github.com/appwrite/appwrite/blob/080c26e3599adcf603e484ec87eb13c9427939d3/app/controllers/general.php#L259

**Comment by JakeAi** [comment]
> @JakeAi Not sure, but there is a comment about the root user in the code,
> 
> https://github.com/appwrite/appwrite/blob/080c26e3599adcf603e484ec87eb13c9427939d3/app/controllers/general.php#L259

Interesting… I guess we will wait for a dev response

**Comment by PineappleIOnic** [comment]
Hm, this could be a root user issue. I'm not sure why though since root would theoretically have full access and should have full control. I'll bring it up with the team.

**Comment by eldadfux** [comment]
> > @JakeAi Not sure, but there is a comment about the root user in the code,
> > https://github.com/appwrite/appwrite/blob/080c26e3599adcf603e484ec87eb13c9427939d3/app/controllers/general.php#L259
> 
> Interesting… I guess we will wait for a dev response

No this should not be related, the missing user attributes in the db might be a hint. When migrating from 0.7 to 0.8 the migrate command should add the new attributes (`sessions`, `tokens`)

**Comment by JakeAi** [comment]
I started on 0.8.0. I never had to migrate.

**Comment by TorstenDittmann** [comment]
This should be fixed now with #1408 🙌🏻
